File: FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/fragment/PreviewImportPgpContactFragment.java
/*
 * © 2016-2018 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.ui.activity.fragment;

import android.app.Activity;
import android.content.OperationApplicationException;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Parcelable;
import android.os.RemoteException;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.flowcrypt.email.R;
import com.flowcrypt.email.database.dao.source.ContactsDaoSource;
import com.flowcrypt.email.js.MessageBlock;
import com.flowcrypt.email.js.PgpContact;
import com.flowcrypt.email.js.PgpKey;
import com.flowcrypt.email.js.core.Js;
import com.flowcrypt.email.model.PublicKeyInfo;
import com.flowcrypt.email.model.results.LoaderResult;
import com.flowcrypt.email.ui.activity.fragment.base.BaseFragment;
import com.flowcrypt.email.ui.adapter.ImportPgpContactsRecyclerViewAdapter;
import com.flowcrypt.email.util.GeneralUtil;
import com.flowcrypt.email.util.UIUtil;
import com.flowcrypt.email.util.exception.ExceptionUtil;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

/**
 * This fragment displays information about public keys owners and information about keys.
 *
 * @author Denis Bondarenko
 * Date: 05.06.2018
 * Time: 14:15
 * E-mail: DenBond7@gmail.com
 */
public class PreviewImportPgpContactFragment extends BaseFragment implements View.OnClickListener {
  private static final String KEY_EXTRA_PUBLIC_KEY_STRING = GeneralUtil.generateUniqueExtraKey
      ("KEY_EXTRA_PUBLIC_KEY_STRING", PreviewImportPgpContactFragment.class);

  private static final String KEY_EXTRA_PUBLIC_KEYS_FILE_URI
      = GeneralUtil.generateUniqueExtraKey("KEY_EXTRA_PUBLIC_KEYS_FILE_URI", PreviewImportPgpContactFragment.class);

  private List<PublicKeyInfo> publicKeyInfoList;
  private RecyclerView recyclerView;
  private TextView btnImportAll;
  private TextView textViewProgressTitle;
  private ProgressBar progressBar;
  private View layoutContentView;
  private View layoutProgress;
  private View emptyView;

  private String publicKeysString;
  private Uri publicKeysFileUri;

  private boolean isParsingStarted;

  public static PreviewImportPgpContactFragment newInstance(String stringExtra, Parcelable parcelableExtra) {
    Bundle args = new Bundle();
    args.putString(KEY_EXTRA_PUBLIC_KEY_STRING, stringExtra);
    args.putParcelable(KEY_EXTRA_PUBLIC_KEYS_FILE_URI, parcelableExtra);

    PreviewImportPgpContactFragment previewImportPgpContactFragment = new PreviewImportPgpContactFragment();
    previewImportPgpContactFragment.setArguments(args);
    return previewImportPgpContactFragment;
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setRetainInstance(true);

    Bundle bundle = getArguments();

    if (bundle != null) {
      publicKeysString = bundle.getString(KEY_EXTRA_PUBLIC_KEY_STRING);
      publicKeysFileUri = bundle.getParcelable(KEY_EXTRA_PUBLIC_KEYS_FILE_URI);
    }
  }

  @Nullable
  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                           @Nullable Bundle savedInstanceState) {
    return inflater.inflate(R.layout.fragment_preview_import_pgp_contact, container, false);
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    initViews(view);
  }

  @Override
  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
    super.onActivityCreated(savedInstanceState);
    if (TextUtils.isEmpty(publicKeysString) && publicKeysFileUri == null) {
      if (getActivity() != null) {
        getActivity().setResult(Activity.RESULT_CANCELED);
        getActivity().finish();
      }
    } else if (!isParsingStarted) {
      new PublicKeysParserAsyncTask(this, publicKeysString, publicKeysFileUri).execute();
      isParsingStarted = true;
    }
  }

  @Override
  public void onClick(View v) {
    switch (v.getId()) {
      case R.id.buttonImportAll:
        new SaveAllContactsAsyncTask(this, publicKeyInfoList).execute();
        break;
    }
  }

  @SuppressWarnings("unchecked")
  @Override
  public void onSuccess(int loaderId, Object result) {
    switch (loaderId) {
      case R.id.loader_id_parse_public_keys:
        publicKeyInfoList = (List<PublicKeyInfo>) result;
        if (!publicKeyInfoList.isEmpty()) {
          UIUtil.exchangeViewVisibility(getContext(), false, layoutProgress, layoutContentView);
          recyclerView.setAdapter(new ImportPgpContactsRecyclerViewAdapter(publicKeyInfoList));
          btnImportAll.setVisibility(publicKeyInfoList.size() > 1 ? View.VISIBLE : View.GONE);
        } else {
          UIUtil.exchangeViewVisibility(getContext(), false, layoutProgress, emptyView);
        }
        break;

      default:
        super.onSuccess(loaderId, result);
    }
  }

  @Override
  public void onError(int loaderId, Exception e) {
    switch (loaderId) {
      case R.id.loader_id_parse_public_keys:
        if (getActivity() != null) {
          getActivity().setResult(Activity.RESULT_CANCELED);
          Toast.makeText(getContext(), TextUtils.isEmpty(e.getMessage()) ? getString(R.string.unknown_error) :
              e.getMessage(), Toast.LENGTH_SHORT).show();
          getActivity().finish();
        }
        break;

      default:
        super.onError(loaderId, e);
    }
  }

  private void initViews(View root) {
    layoutContentView = root.findViewById(R.id.layoutContentView);
    layoutProgress = root.findViewById(R.id.layoutProgress);
    btnImportAll = root.findViewById(R.id.buttonImportAll);
    textViewProgressTitle = root.findViewById(R.id.textViewProgressTitle);
    progressBar = root.findViewById(R.id.progressBar);
    btnImportAll.setOnClickListener(this);
    recyclerView = root.findViewById(R.id.recyclerViewContacts);
    recyclerView.setHasFixedSize(true);
    recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    emptyView = root.findViewById(R.id.emptyView);
  }

  private void handleImportAllResult(Boolean result) {
    if (isAdded()) {
      if (result) {
        Toast.makeText(getContext(), R.string.success, Toast.LENGTH_SHORT).show();
        if (getActivity() != null) {
          getActivity().setResult(Activity.RESULT_OK);
          getActivity().finish();
        }
      } else {
        UIUtil.exchangeViewVisibility(getContext(), false, layoutProgress, layoutContentView);
        Toast.makeText(getContext(), R.string.unknown_error, Toast.LENGTH_SHORT).show();
      }
    }
  }

  private static class PublicKeysParserAsyncTask extends BaseAsyncTask<Void, Integer, LoaderResult> {
    private final String publicKeysString;
    private final Uri publicKeysFileUri;

    PublicKeysParserAsyncTask(PreviewImportPgpContactFragment fragment, String publicKeysString, Uri uri) {
      super(fragment);
      this.publicKeysString = publicKeysString;
      this.publicKeysFileUri = uri;
    }

    @Override
    protected LoaderResult doInBackground(Void... uris) {
      String armoredKeys = publicKeysString;

      try {
        if (publicKeysFileUri != null && weakRef.get() != null) {
          armoredKeys = GeneralUtil.readFileFromUriToString(weakRef.get().getContext(), publicKeysFileUri);
        }
      } catch (IOException e) {
        e.printStackTrace();
        return new LoaderResult(null, e);
      }

      if (!TextUtils.isEmpty(armoredKeys) && weakRef.get() != null) {
        return parseKeys(armoredKeys);
      } else {
        return new LoaderResult(null, new NullPointerException("An input string is null!"));
      }
    }

    @Override
    protected void onPostExecute(LoaderResult loaderResult) {
      super.onPostExecute(loaderResult);
      if (weakRef.get() != null) {
        weakRef.get().handleLoaderResult(R.id.loader_id_parse_public_keys, loaderResult);
      }
    }

    @Override
    public int getProgressTitleResourcesId() {
      return R.string.parsing_public_keys;
    }

    @Override
    public void updateProgress(Integer integer) {
      if (weakRef.get() != null) {
        weakRef.get().progressBar.setProgress(integer);
      }
    }

    private LoaderResult parseKeys(String armoredKeys) {
      try {
        Js js = new Js(weakRef.get().getContext(), null);
        String normalizedArmoredKey = js.crypto_key_normalize(armoredKeys);
        PgpKey pgpKey = js.crypto_key_read(normalizedArmoredKey);

        if (js.is_valid_key(pgpKey, false)) {
          return new LoaderResult(parsePublicKeysInfo(js, armoredKeys), null);
        } else {
          if (weakRef.get() != null) {
            return new LoaderResult(null, new IllegalArgumentException(
                weakRef.get().getContext().getString(R.string.clipboard_has_wrong_structure,
                    weakRef.get().getContext().getString(R.string.public_))));
          } else {
            return new LoaderResult(null,
                new IllegalArgumentException("The content of your clipboard doesn't look like a valid PGP pubkey."));
          }
        }

      } catch (Exception e) {
        e.printStackTrace();
        ExceptionUtil.handleError(e);
        return new LoaderResult(null, e);
      }
    }

    private List<PublicKeyInfo> parsePublicKeysInfo(Js js, @NonNull String armoredKeys) {
      List<PublicKeyInfo> publicKeyInfoList = new ArrayList<>();

      if (TextUtils.isEmpty(armoredKeys)) {
        return publicKeyInfoList;
      }

      Set<String> emails = new HashSet<>();
      MessageBlock[] msgBlocks = js.crypto_armor_detect_blocks(armoredKeys);

      int blocksCount = msgBlocks.length;
      float progress;
      float lastProgress = 0;

      for (int i = 0; i < msgBlocks.length; i++) {
        MessageBlock messageBlock = msgBlocks[i];
        if (messageBlock != null && messageBlock.getType() != null) {
          switch (messageBlock.getType()) {
            case MessageBlock.TYPE_PGP_PUBLIC_KEY:
              PublicKeyInfo publicKeyInfo = getPublicKeyInfo(js, emails, messageBlock);

              if (publicKeyInfo != null) {
                publicKeyInfoList.add(publicKeyInfo);
              }
              break;
          }
        }

        progress = i * 100f / blocksCount;
        if (progress - lastProgress >= 1) {
          publishProgress((int) progress);
          lastProgress = progress;
        }
      }

      publishProgress(100);

      return publicKeyInfoList;
    }

    private PublicKeyInfo getPublicKeyInfo(Js js, Set<String> emails, MessageBlock messageBlock) {
      String content = messageBlock.getContent();
      String fingerprint = js.crypto_key_fingerprint(js.crypto_key_read(content));
      String longId = js.crypto_key_longid(fingerprint);
      String keyWords = js.mnemonic(longId);
      PgpKey pgpKey = js.crypto_key_read(content);
      String keyOwner = pgpKey.getPrimaryUserId().getEmail();

      if (keyOwner != null) {
        keyOwner = keyOwner.toLowerCase();

        if (emails.contains(keyOwner)) {
          return null;
        }

        emails.add(keyOwner);

        if (weakRef.get() != null) {
          PgpContact contact = new ContactsDaoSource().getPgpContact(weakRef.get().getContext(), keyOwner);
          return new PublicKeyInfo(keyWords, fingerprint, keyOwner, longId, contact, content);
        }
      }
      return null;
    }
  }

  private static class SaveAllContactsAsyncTask extends BaseAsyncTask<Void, Integer, Boolean> {
    private static final int STEP_AMOUNT = 50;

    private final List<PublicKeyInfo> publicKeyInfoList;

    SaveAllContactsAsyncTask(PreviewImportPgpContactFragment fragment, List<PublicKeyInfo> publicKeyInfoList) {
      super(fragment);
      this.publicKeyInfoList = publicKeyInfoList;
    }

    @Override
    protected Boolean doInBackground(Void... uris) {
      ContactsDaoSource source = new ContactsDaoSource();
      List<PgpContact> newCandidates = new ArrayList<>();
      List<PgpContact> updateCandidates = new ArrayList<>();

      for (PublicKeyInfo publicKeyInfo : publicKeyInfoList) {
        PgpContact pgpContact = new PgpContact(publicKeyInfo.getKeyOwner(), null, publicKeyInfo.getPublicKey(),
            true, null, false, publicKeyInfo.getFingerprint(), publicKeyInfo.getLongId(),
            publicKeyInfo.getKeyWords(), 0);

        if (publicKeyInfo.hasPgpContact()) {
          if (publicKeyInfo.isUpdateEnabled()) {
            updateCandidates.add(pgpContact);
          }
        } else {
          newCandidates.add(pgpContact);
        }
      }

      try {
        float progress;
        float lastProgress = 0f;
        int totalOperationsCount = newCandidates.size() + updateCandidates.size();

        for (int i = 0; i < newCandidates.size(); i++) {
          int start = i;
          int end = newCandidates.size() - i > STEP_AMOUNT ? i + STEP_AMOUNT : newCandidates.size();

          if (weakRef.get() != null) {
            source.addRowsUsingApplyBatch(weakRef.get().getContext(), newCandidates.subList(start, end));
          }
          i = end;

          progress = i * 100f / totalOperationsCount;
          if (progress - lastProgress >= 1) {
            publishProgress((int) progress);
            lastProgress = progress;
          }

          i--;
        }

        for (int i = 0; i < updateCandidates.size(); i++) {
          int start = i;
          int end = updateCandidates.size() - i > STEP_AMOUNT ? i + STEP_AMOUNT : updateCandidates.size() - 1;

          if (weakRef.get() != null) {
            source.updatePgpContacts(weakRef.get().getContext(), updateCandidates.subList(start, end + 1));
          }
          i = end + 1;

          progress = i * 100f / totalOperationsCount;
          if (progress - lastProgress >= 1) {
            publishProgress((int) progress);
            lastProgress = progress;
          }
        }

      } catch (RemoteException | OperationApplicationException e) {
        e.printStackTrace();
        return false;
      }

      publishProgress(100);
      return true;
    }


    @Override
    protected void onPostExecute(Boolean b) {
      super.onPostExecute(b);
      if (b != null && weakRef.get() != null) {
        weakRef.get().handleImportAllResult(b);
      }
    }

    @Override
    public int getProgressTitleResourcesId() {
      return R.string.importing_public_keys;
    }

    @Override
    public void updateProgress(Integer integer) {
      if (weakRef.get() != null) {
        weakRef.get().progressBar.setProgress(integer);
      }
    }
  }

  private abstract static class BaseAsyncTask<Params, Progress, Result> extends AsyncTask<Params, Progress, Result> {
    final WeakReference<PreviewImportPgpContactFragment> weakRef;

    BaseAsyncTask(PreviewImportPgpContactFragment previewImportPgpContactFragment) {
      this.weakRef = new WeakReference<>(previewImportPgpContactFragment);
    }

    public abstract int getProgressTitleResourcesId();

    public abstract void updateProgress(Progress progress);

    @Override
    protected void onPreExecute() {
      super.onPreExecute();
      if (weakRef.get() != null) {
        weakRef.get().progressBar.setIndeterminate(true);
        weakRef.get().textViewProgressTitle.setText(getProgressTitleResourcesId());
        UIUtil.exchangeViewVisibility(weakRef.get().getContext(), true,
            weakRef.get().layoutProgress, weakRef.get().layoutContentView);
      }
    }

    @SafeVarargs
    @Override
    protected final void onProgressUpdate(Progress... values) {
      super.onProgressUpdate(values);
      if (weakRef.get() != null) {
        if (weakRef.get().progressBar.isIndeterminate()) {
          weakRef.get().progressBar.setIndeterminate(false);
        }
        updateProgress(values[0]);
      }
    }
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/fragment/PreviewImportPgpContactFragment.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors:
 *   DenBond7
 *   Ivan Pizhenko
 */

package com.flowcrypt.email.ui.activity.fragment

import android.app.Activity
import android.content.OperationApplicationException
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.os.Parcelable
import android.os.RemoteException
import android.text.TextUtils
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.flowcrypt.email.Constants
import com.flowcrypt.email.R
import com.flowcrypt.email.database.FlowCryptRoomDatabase
import com.flowcrypt.email.database.entity.ContactEntity
import com.flowcrypt.email.jetpack.viewmodel.ContactsViewModel
import com.flowcrypt.email.model.PgpContact
import com.flowcrypt.email.model.PublicKeyInfo
import com.flowcrypt.email.model.results.LoaderResult
import com.flowcrypt.email.security.model.PgpKeyDetails
import com.flowcrypt.email.security.pgp.PgpKey
import com.flowcrypt.email.ui.activity.fragment.base.BaseFragment
import com.flowcrypt.email.ui.adapter.ImportPgpContactsRecyclerViewAdapter
import com.flowcrypt.email.util.GeneralUtil
import com.flowcrypt.email.util.UIUtil
import com.flowcrypt.email.util.exception.ExceptionUtil
import com.google.android.gms.common.util.CollectionUtils
import java.io.IOException
import java.io.InputStream
import java.lang.ref.WeakReference
import java.util.*

/**
 * This fragment displays information about public keys owners and information about keys.
 *
 * @author Denis Bondarenko
 * Date: 05.06.2018
 * Time: 14:15
 * E-mail: DenBond7@gmail.com
 */
//todo-denbond7 Improve this class.
//Need to migrate to use LiveData and improve performance for the file parsing.
class PreviewImportPgpContactFragment : BaseFragment(), View.OnClickListener,
    ImportPgpContactsRecyclerViewAdapter.ContactActionsListener {

  private var recyclerView: RecyclerView? = null
  private var btnImportAll: TextView? = null
  private var textViewProgressTitle: TextView? = null
  private var progressBar: ProgressBar? = null
  private var layoutContentView: View? = null
  private var layoutProgress: View? = null
  private var emptyView: View? = null

  private var publicKeysString: String? = null
  private var publicKeysFileUri: Uri? = null

  private val contactsViewModel: ContactsViewModel by viewModels()
  private val adapter: ImportPgpContactsRecyclerViewAdapter = ImportPgpContactsRecyclerViewAdapter()
  private var isParsingStarted: Boolean = false

  override val contentResourceId: Int = R.layout.fragment_preview_import_pgp_contact

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    retainInstance = true

    val bundle = arguments

    if (bundle != null) {
      publicKeysString = bundle.getString(KEY_EXTRA_PUBLIC_KEY_STRING)
      publicKeysFileUri = bundle.getParcelable(KEY_EXTRA_PUBLIC_KEYS_FILE_URI)
    }

    adapter.contactActionsListener = this
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    initViews(view)
  }

  override fun onActivityCreated(savedInstanceState: Bundle?) {
    super.onActivityCreated(savedInstanceState)
    if (TextUtils.isEmpty(publicKeysString) && publicKeysFileUri == null) {
      if (activity != null) {
        requireActivity().setResult(Activity.RESULT_CANCELED)
        requireActivity().finish()
      }
    } else if (!isParsingStarted) {
      PublicKeysParserAsyncTask(this, publicKeysString ?: "", publicKeysFileUri).execute()
      isParsingStarted = true
    }
  }

  override fun onClick(v: View) {
    when (v.id) {
      R.id.buttonImportAll -> SaveAllContactsAsyncTask(this, adapter.publicKeys).execute()
    }
  }

  @Suppress("UNCHECKED_CAST")
  override fun onSuccess(loaderId: Int, result: Any?) {
    when (loaderId) {
      R.id.loader_id_parse_public_keys -> {
        val list = result as? List<PublicKeyInfo> ?: emptyList()
        if (list.isNotEmpty()) {
          UIUtil.exchangeViewVisibility(false, layoutProgress, layoutContentView)
          adapter.swap(list)
          btnImportAll?.visibility = if (list.size > 1) View.VISIBLE else View.GONE
        } else {
          UIUtil.exchangeViewVisibility(false, layoutProgress, emptyView)
        }
      }

      else -> super.onSuccess(loaderId, result)
    }
  }

  override fun onError(loaderId: Int, e: Exception?) {
    when (loaderId) {
      R.id.loader_id_parse_public_keys -> if (activity != null) {
        requireActivity().setResult(Activity.RESULT_CANCELED)
        Toast.makeText(context, if (e?.message.isNullOrEmpty()) {
          e?.javaClass?.simpleName ?: getString(R.string.unknown_error)
        } else
          e?.message, Toast.LENGTH_SHORT).show()
        requireActivity().finish()
      }

      else -> super.onError(loaderId, e)
    }
  }

  override fun onSaveContactClick(publicKeyInfo: PublicKeyInfo) {
    contactsViewModel.addContact(publicKeyInfo.toPgpContact())
  }

  override fun onUpdateContactClick(publicKeyInfo: PublicKeyInfo) {
    contactsViewModel.updateContact(publicKeyInfo.toPgpContact())
  }

  private fun initViews(root: View) {
    layoutContentView = root.findViewById(R.id.layoutContentView)
    layoutProgress = root.findViewById(R.id.layoutProgress)
    btnImportAll = root.findViewById(R.id.buttonImportAll)
    textViewProgressTitle = root.findViewById(R.id.textViewProgressTitle)
    progressBar = root.findViewById(R.id.progressBar)
    btnImportAll?.setOnClickListener(this)
    recyclerView = root.findViewById(R.id.recyclerViewContacts)
    recyclerView?.setHasFixedSize(true)
    recyclerView?.layoutManager = LinearLayoutManager(context)
    recyclerView?.adapter = adapter
    emptyView = root.findViewById(R.id.emptyView)
  }

  private fun handleImportAllResult(result: Boolean?) {
    if (isAdded) {
      if (result == true) {
        Toast.makeText(context, R.string.success, Toast.LENGTH_SHORT).show()
        if (activity != null) {
          requireActivity().setResult(Activity.RESULT_OK)
          requireActivity().finish()
        }
      } else {
        UIUtil.exchangeViewVisibility(false, layoutProgress, layoutContentView)
        Toast.makeText(context, getString(R.string.could_not_import_data), Toast.LENGTH_SHORT).show()
      }
    }
  }

  private class PublicKeysParserAsyncTask(fragment: PreviewImportPgpContactFragment,
                                          private val publicKeysString: String,
                                          private val publicKeysFileUri: Uri?)
    : BaseAsyncTask<Void, Int, LoaderResult>(fragment) {
    override val progressTitleResourcesId: Int
      get() = R.string.parsing_public_keys

    override fun doInBackground(vararg uris: Void): LoaderResult {
      var keysInputStream: InputStream? = publicKeysString.toByteArray().inputStream()

      try {
        if (publicKeysFileUri != null && weakRef.get() != null) {
          weakRef.get()?.context?.let {
            keysInputStream = it.contentResolver.openInputStream(publicKeysFileUri)
          }
        }
      } catch (e: IOException) {
        e.printStackTrace()
        return LoaderResult(null, e)
      }

      return if (keysInputStream != null) {
        parseKeys(keysInputStream!!)
      } else {
        LoaderResult(null, IllegalStateException("Source is null"))
      }
    }

    override fun onPostExecute(loaderResult: LoaderResult) {
      super.onPostExecute(loaderResult)
      if (weakRef.get() != null) {
        weakRef.get()?.handleLoaderResult(R.id.loader_id_parse_public_keys, loaderResult)
      }
    }

    override fun updateProgress(progress: Int) {
      if (weakRef.get() != null) {
        weakRef.get()?.progressBar!!.progress = progress
      }
    }

    private fun parseKeys(inputStream: InputStream): LoaderResult {
      try {
        val details = PgpKey.parseKeys(inputStream).toNodeKeyDetailsList()

        return if (!CollectionUtils.isEmpty(details)) {
          LoaderResult(parsePublicKeysInfo(details), null)
        } else {
          if (weakRef.get() != null) {
            LoaderResult(null, IllegalArgumentException(
                weakRef.get()?.requireContext()?.getString(R.string.clipboard_has_wrong_structure,
                    weakRef.get()?.requireContext()?.getString(R.string.public_))))
          } else {
            LoaderResult(null,
                IllegalArgumentException("The content of your clipboard doesn't look like a valid PGP pubkey."))
          }
        }

      } catch (e: Exception) {
        e.printStackTrace()
        ExceptionUtil.handleError(e)
        return LoaderResult(null, e)
      }
    }

    private fun parsePublicKeysInfo(details: List<PgpKeyDetails>): List<PublicKeyInfo> {
      val publicKeyInfoList = ArrayList<PublicKeyInfo>()

      val emails = HashSet<String>()

      val blocksCount = details.size
      var progress: Float
      var lastProgress = 0f

      for (i in 0 until blocksCount) {
        val nodeKeyDetails = details[i]
        getPublicKeyInfo(nodeKeyDetails, emails)?.let {
          if (it.publicKey.length <= Constants.MAX_PUB_KEY_SIZE) {
            publicKeyInfoList.add(it)
          }
        }

        progress = i * 100f / blocksCount
        if (progress - lastProgress >= 1) {
          publishProgress(progress.toInt())
          lastProgress = progress
        }
      }

      publishProgress(100)

      return publicKeyInfoList
    }

    private fun getPublicKeyInfo(pgpKeyDetails: PgpKeyDetails, emails: MutableSet<String>): PublicKeyInfo? {
      val fingerprint = pgpKeyDetails.fingerprint
      var keyOwner: String? = pgpKeyDetails.primaryPgpContact.email

      if (keyOwner != null) {
        keyOwner = keyOwner.toLowerCase(Locale.getDefault())

        if (emails.contains(keyOwner)) {
          return null
        }

        emails.add(keyOwner)

        if (weakRef.get() != null) {
          val contact = FlowCryptRoomDatabase.getDatabase(weakRef.get()?.requireContext()!!)
              .contactsDao().getContactByEmail(keyOwner)?.toPgpContact()
          return PublicKeyInfo(fingerprint!!, keyOwner, contact, pgpKeyDetails.publicKey!!)
        }
      }
      return null
    }
  }

  private class SaveAllContactsAsyncTask(
      fragment: PreviewImportPgpContactFragment,
      private val publicKeyInfoList: List<PublicKeyInfo>) : BaseAsyncTask<Void, Int, Boolean>(fragment) {

    override val progressTitleResourcesId: Int
      get() = R.string.importing_public_keys

    override fun doInBackground(vararg uris: Void): Boolean {
      val newCandidates = ArrayList<PgpContact>()
      val updateCandidates = ArrayList<PgpContact>()

      for (publicKeyInfo in publicKeyInfoList) {
        val pgpContact = PgpContact(publicKeyInfo.keyOwner, null, publicKeyInfo.publicKey,
            true, null, publicKeyInfo.fingerprint, 0)

        if (publicKeyInfo.hasPgpContact()) {
          if (publicKeyInfo.isUpdateEnabled) {
            updateCandidates.add(pgpContact)
          }
        } else {
          newCandidates.add(pgpContact)
        }
      }

      try {
        var progress: Float
        var lastProgress = 0f
        val totalOperationsCount = newCandidates.size + updateCandidates.size

        run {
          var i = 0
          while (i < newCandidates.size) {
            val start = i
            val end = if (newCandidates.size - i > STEP_AMOUNT) i + STEP_AMOUNT else newCandidates.size

            if (weakRef.get() != null) {
              FlowCryptRoomDatabase.getDatabase(weakRef.get()?.requireContext()!!)
                  .contactsDao().insert(newCandidates.subList(start, end).map { it.toContactEntity() })
            }
            i = end

            progress = i * 100f / totalOperationsCount
            if (progress - lastProgress >= 1) {
              publishProgress(progress.toInt())
              lastProgress = progress
            }

            i--
            i++
          }
        }

        var i = 0
        while (i < updateCandidates.size) {
          val start = i
          val end = if (updateCandidates.size - i > STEP_AMOUNT) i + STEP_AMOUNT else updateCandidates.size - 1

          if (weakRef.get() != null) {
            val contacts = mutableListOf<ContactEntity>()
            val list = updateCandidates.subList(start, end + 1)

            list.forEach { pgpContact ->
              val foundContactEntity = FlowCryptRoomDatabase.getDatabase(weakRef.get()?.requireContext()!!)
                  .contactsDao().getContactByEmail(pgpContact.email)
              foundContactEntity?.let { entity -> contacts.add(pgpContact.toContactEntity().copy(id = entity.id)) }
            }

            FlowCryptRoomDatabase.getDatabase(weakRef.get()?.requireContext()!!).contactsDao().update(contacts)
          }
          i = end + 1

          progress = i * 100f / totalOperationsCount
          if (progress - lastProgress >= 1) {
            publishProgress(progress.toInt())
            lastProgress = progress
          }
          i++
        }

      } catch (e: RemoteException) {
        e.printStackTrace()
        return false
      } catch (e: OperationApplicationException) {
        e.printStackTrace()
        return false
      }

      publishProgress(100)
      return true
    }


    override fun onPostExecute(b: Boolean?) {
      super.onPostExecute(b)
      if (b != null && weakRef.get() != null) {
        weakRef.get()?.handleImportAllResult(b)
      }
    }

    override fun updateProgress(progress: Int) {
      if (weakRef.get() != null) {
        weakRef.get()?.progressBar!!.progress = progress
      }
    }

    companion object {
      private const val STEP_AMOUNT = 50
    }
  }

  private abstract class BaseAsyncTask<Params, Progress, Result>(previewImportPgpContactFragment: PreviewImportPgpContactFragment)
    : AsyncTask<Params, Progress, Result>() {
    val weakRef: WeakReference<PreviewImportPgpContactFragment> =
        WeakReference(previewImportPgpContactFragment)

    abstract val progressTitleResourcesId: Int

    abstract fun updateProgress(progress: Progress)

    override fun onPreExecute() {
      super.onPreExecute()
      if (weakRef.get() != null) {
        weakRef.get()?.progressBar!!.isIndeterminate = true
        weakRef.get()?.textViewProgressTitle!!.setText(progressTitleResourcesId)
        UIUtil.exchangeViewVisibility(true,
            weakRef.get()?.layoutProgress!!, weakRef.get()?.layoutContentView!!)
      }
    }

    @SafeVarargs
    override fun onProgressUpdate(vararg values: Progress) {
      super.onProgressUpdate(*values)
      if (weakRef.get() != null) {
        if (weakRef.get()?.progressBar!!.isIndeterminate) {
          weakRef.get()?.progressBar!!.isIndeterminate = false
        }
        updateProgress(values[0])
      }
    }
  }

  companion object {
    private val KEY_EXTRA_PUBLIC_KEY_STRING =
        GeneralUtil.generateUniqueExtraKey("KEY_EXTRA_PUBLIC_KEY_STRING",
            PreviewImportPgpContactFragment::class.java)

    private val KEY_EXTRA_PUBLIC_KEYS_FILE_URI =
        GeneralUtil.generateUniqueExtraKey("KEY_EXTRA_PUBLIC_KEYS_FILE_URI",
            PreviewImportPgpContactFragment::class.java)

    fun newInstance(stringExtra: String?, fileUri: Parcelable?): PreviewImportPgpContactFragment {
      val args = Bundle()
      args.putString(KEY_EXTRA_PUBLIC_KEY_STRING, stringExtra)
      args.putParcelable(KEY_EXTRA_PUBLIC_KEYS_FILE_URI, fileUri)

      val previewImportPgpContactFragment = PreviewImportPgpContactFragment()
      previewImportPgpContactFragment.arguments = args
      return previewImportPgpContactFragment
    }
  }
}
